File: backend/functions/src/qdrant.ts
const qdrantUrl = process.env.QDRANT_URL || "http://127.0.0.1:6333";

// Use a versioned default collection name so the new dense+sparse schema does not
// collide with older single-vector collections that may already exist locally.
export const vectorCollectionName = process.env.QDRANT_COLLECTION || "pocketpilot_vectors_v2";
export const denseVectorName = "dense";
export const sparseVectorName = "sparse";
export const ragVectorSource = "rag";

const QDRANT_TIMEOUT_MS = 30_000;
const SPARSE_HASH_SPACE = 262_139;
const RRF_K = 60;

export interface SparseVector {
  indices: number[];
  values: number[];
}

export interface SparseBoost {
  token: string;
  weight: number;
}

export interface PayloadFilterCondition {
  key: string;
  match?: {
    value?: string | number | boolean;
    any?: Array<string | number | boolean>;
  };
  range?: {
    gte?: number;
    lte?: number;
    gt?: number;
    lt?: number;
  };
}

export interface VectorPayloadRecord extends Record<string, unknown> {
  userId: string;
  kind: string;
  refId: string;
  text: string;
  tags?: string[];
  source: string;
  contentHash?: string;
  createdAt: string;
  updatedAt?: string;
}

interface QdrantPoint {
  id: string | number;
  score?: number;
  payload?: VectorPayloadRecord;
}

interface QdrantQueryResult {
  points?: QdrantPoint[];
}

interface QdrantScrollResult {
  points?: QdrantPoint[];
  next_page_offset?: string | number | null;
}

function qdrantHeaders(): Record<string, string> {
  return {
    "Content-Type": "application/json",
  };
}

async function qdrantRequest<T>(path: string, init: RequestInit = {}): Promise<T> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), QDRANT_TIMEOUT_MS);

  try {
    const response = await fetch(`${qdrantUrl}${path}`, {
      ...init,
      headers: {
        ...qdrantHeaders(),
        ...(init.headers || {}),
      },
      signal: controller.signal,
    });

    if (response.status === 404) {
      throw new Error("QDRANT_NOT_FOUND");
    }

    const json = (await response.json().catch(() => ({}))) as {
      status?: string;
      result?: T;
      error?: string;
    };

    if (!response.ok) {
      throw new Error(json.error || `Qdrant request failed with ${response.status}`);
    }

    return json.result as T;
  } finally {
    clearTimeout(timeout);
  }
}

function createFilter(
  userId: string,
  kinds?: string[],
  extraMust: PayloadFilterCondition[] = []
): { must: Array<Record<string, unknown> | PayloadFilterCondition> } {
  const must: Array<Record<string, unknown> | PayloadFilterCondition> = [
    {
      key: "userId",
      match: { value: userId },
    },
    {
      key: "source",
      match: { value: ragVectorSource },
    },
  ];

  if (kinds && kinds.length > 0) {
    must.push({
      key: "kind",
      match: { any: kinds },
    });
  }

  must.push(...extraMust);

  return { must };
}

function fnv1a32(input: string): number {
  let hash = 0x811c9dc5;
  for (let index = 0; index < input.length; index += 1) {
    hash ^= input.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

function tokenizeForSparseIndex(text: string): Array<{ token: string; weight: number }> {
  const lowered = text.toLowerCase();
  const entries = new Map<string, number>();

  const addToken = (token: string, weight = 1) => {
    const normalized = token.trim().toLowerCase();
    if (!normalized) return;
    entries.set(normalized, (entries.get(normalized) || 0) + weight);
  };

  const dateMatches =
    lowered.match(/\b\d{4}-\d{2}(?:-\d{2})?\b/g) ||
    lowered.match(/\b(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*\s+\d{4}\b/g) ||
    [];
  for (const match of dateMatches) {
    addToken(match, 3);
  }

  const moneyMatches = lowered.match(/\$?\d+(?:\.\d+)?/g) || [];
  for (const match of moneyMatches) {
    const normalized = match.replace(/^\$/, "");
    const weight =
      normalized.includes(".") || normalized.length >= 4 || match.includes("$") ? 2.5 : 1.2;
    addToken(normalized, weight);
    if (match.includes("$")) {
      addToken(match, weight);
    }
  }

  const wordMatches = lowered.match(/[a-z0-9][a-z0-9._:/-]{1,}/g) || [];
  for (const match of wordMatches) {
    if (match.length < 2) continue;
    const hasDigit = /\d/.test(match);
    addToken(match, hasDigit ? 1.8 : 1);
  }

  return Array.from(entries.entries()).map(([token, weight]) => ({ token, weight }));
}

export function buildSparseVector(text: string, boosts: SparseBoost[] = []): SparseVector {
  const weightedTokens = tokenizeForSparseIndex(text);
  const indicesToValues = new Map<number, number>();

  for (const { token, weight } of weightedTokens) {
    const index = fnv1a32(token) % SPARSE_HASH_SPACE;
    indicesToValues.set(index, (indicesToValues.get(index) || 0) + weight);
  }

  for (const { token, weight } of boosts) {
    const normalized = token.trim().toLowerCase();
    if (!normalized || weight <= 0) continue;
    const index = fnv1a32(normalized) % SPARSE_HASH_SPACE;
    indicesToValues.set(index, (indicesToValues.get(index) || 0) + weight);
  }

  const ordered = Array.from(indicesToValues.entries()).sort((a, b) => a[0] - b[0]);
  return {
    indices: ordered.map(([index]) => index),
    values: ordered.map(([, value]) => Number(value.toFixed(4))),
  };
}

function hasExpectedCollectionSchema(
  collection: {
    config?: {
      params?: {
        vectors?: Record<string, { size?: number }>;
        sparse_vectors?: Record<string, unknown>;
      };
    };
  } | null,
  dimension: number
): boolean {
  const vectors = collection?.config?.params?.vectors;
  const sparseVectors = collection?.config?.params?.sparse_vectors;

  if (!vectors || typeof vectors !== "object") {
    return false;
  }

  const denseConfig = vectors[denseVectorName];
  if (!denseConfig || denseConfig.size !== dimension) {
    return false;
  }

  return !!sparseVectors && typeof sparseVectors === "object" && sparseVectorName in sparseVectors;
}

async function createCollection(dimension: number): Promise<void> {
  await qdrantRequest(`/collections/${vectorCollectionName}`, {
    method: "PUT",
    body: JSON.stringify({
      vectors: {
        [denseVectorName]: {
          size: dimension,
          distance: "Cosine",
        },
      },
      sparse_vectors: {
        [sparseVectorName]: {},
      },
    }),
  });
}

export async function ensureVectorCollection(dimension: number): Promise<void> {
  let existing: {
    config?: {
      params?: {
        vectors?: Record<string, { size?: number }>;
        sparse_vectors?: Record<string, unknown>;
      };
    };
  } | null = null;

  try {
    existing = await qdrantRequest(`/collections/${vectorCollectionName}`, {
      method: "GET",
    });
  } catch (error) {
    if (!(error instanceof Error) || error.message !== "QDRANT_NOT_FOUND") {
      throw error;
    }
  }

  if (!existing) {
    await createCollection(dimension);
    return;
  }

  if (hasExpectedCollectionSchema(existing, dimension)) {
    return;
  }

  await qdrantRequest(`/collections/${vectorCollectionName}`, {
    method: "DELETE",
  });
  await createCollection(dimension);
}

export async function upsertPoints(
  points: Array<{
    id: string;
    denseVector: number[];
    sparseVector: SparseVector;
    payload: VectorPayloadRecord;
  }>
): Promise<void> {
  if (points.length === 0) {
    return;
  }

  await qdrantRequest(`/collections/${vectorCollectionName}/points?wait=true`, {
    method: "PUT",
    body: JSON.stringify({
      points: points.map((point) => ({
        id: point.id,
        vector: {
          [denseVectorName]: point.denseVector,
          [sparseVectorName]: point.sparseVector,
        },
        payload: point.payload,
      })),
    }),
  });
}

export async function deletePoints(ids: Array<string | number>): Promise<void> {
  if (ids.length === 0) {
    return;
  }

  await qdrantRequest(`/collections/${vectorCollectionName}/points/delete?wait=true`, {
    method: "POST",
    body: JSON.stringify({
      points: ids,
    }),
  });
}

async function queryPoints(params: {
  query: number[] | SparseVector;
  using: string;
  userId: string;
  kinds?: string[];
  extraMust?: PayloadFilterCondition[];
  limit: number;
  scoreThreshold?: number;
}): Promise<QdrantPoint[]> {
  const result = await qdrantRequest<QdrantQueryResult>(
    `/collections/${vectorCollectionName}/points/query`,
    {
      method: "POST",
      body: JSON.stringify({
        query: params.query,
        using: params.using,
        limit: params.limit,
        score_threshold: params.scoreThreshold,
        filter: createFilter(params.userId, params.kinds, params.extraMust),
        with_payload: true,
        with_vector: false,
      }),
    }
  );

  return result.points || [];
}

export async function queryDensePoints(params: {
  vector: number[];
  userId: string;
  kinds?: string[];
  extraMust?: PayloadFilterCondition[];
  limit: number;
  scoreThreshold?: number;
}): Promise<QdrantPoint[]> {
  return queryPoints({
    query: params.vector,
    using: denseVectorName,
    userId: params.userId,
    kinds: params.kinds,
    extraMust: params.extraMust,
    limit: params.limit,
    scoreThreshold: params.scoreThreshold,
  });
}

function fuseReciprocalRank(
  dense: QdrantPoint[],
  sparse: QdrantPoint[],
  preferSparse: boolean,
  limit: number
): QdrantPoint[] {
  const denseWeight = preferSparse ? 0.4 : 0.6;
  const sparseWeight = preferSparse ? 0.6 : 0.4;
  const merged = new Map<string, QdrantPoint & { fusedScore: number }>();

  const addRanked = (points: QdrantPoint[], weight: number) => {
    points.forEach((point, index) => {
      const key = String(point.id);
      const current = merged.get(key) || { ...point, fusedScore: 0 };
      current.fusedScore += weight / (RRF_K + index + 1);
      if (!current.payload && point.payload) {
        current.payload = point.payload;
      }
      merged.set(key, current);
    });
  };

  addRanked(dense, denseWeight);
  addRanked(sparse, sparseWeight);

  return Array.from(merged.values())
    .sort((a, b) => b.fusedScore - a.fusedScore)
    .slice(0, limit)
    .map((point) => ({
      id: point.id,
      payload: point.payload,
      score: point.fusedScore,
    }));
}

export async function searchHybridPoints(params: {
  denseVector: number[];
  sparseVector: SparseVector;
  userId: string;
  kinds?: string[];
  extraMust?: PayloadFilterCondition[];
  limit: number;
  preferSparse: boolean;
  scoreThreshold?: number;
}): Promise<QdrantPoint[]> {
  const candidateLimit = Math.min(Math.max(params.limit * 3, 12), 60);

  const [denseResults, sparseResults] = await Promise.all([
    queryDensePoints({
      vector: params.denseVector,
      userId: params.userId,
      kinds: params.kinds,
      extraMust: params.extraMust,
      limit: candidateLimit,
      scoreThreshold: params.scoreThreshold,
    }),
    params.sparseVector.indices.length > 0
      ? queryPoints({
          query: params.sparseVector,
          using: sparseVectorName,
          userId: params.userId,
          kinds: params.kinds,
          extraMust: params.extraMust,
          limit: candidateLimit,
        })
      : Promise.resolve([]),
  ]);

  if (sparseResults.length === 0) {
    return denseResults.slice(0, params.limit);
  }

  return fuseReciprocalRank(denseResults, sparseResults, params.preferSparse, params.limit);
}

export async function scrollUserSourcePoints(userId: string): Promise<QdrantPoint[]> {
  const points: QdrantPoint[] = [];
  let offset: string | number | null | undefined;

  while (true) {
    const result = await qdrantRequest<QdrantScrollResult>(
      `/collections/${vectorCollectionName}/points/scroll`,
      {
        method: "POST",
        body: JSON.stringify({
          offset,
          limit: 256,
          filter: createFilter(userId),
          with_payload: true,
          with_vector: false,
        }),
      }
    );

    const page = result.points || [];
    points.push(...page);

    if (!result.next_page_offset || page.length === 0) {
      break;
    }

    offset = result.next_page_offset;
  }

  return points;
}

File: backend/rust/crates/common/src/qdrant.rs
use crate::{config::AppConfig, error::AppError, error::AppResult};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::HashMap;

pub const DENSE_VECTOR_NAME: &str = "dense";
pub const SPARSE_VECTOR_NAME: &str = "sparse";
pub const RAG_VECTOR_SOURCE: &str = "rag";
const SPARSE_HASH_SPACE: u32 = 262_139;
const RRF_K: f32 = 60.0;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SparseVector {
    pub indices: Vec<u32>,
    pub values: Vec<f32>,
}

#[derive(Debug, Clone)]
pub struct SparseBoost {
    pub token: String,
    pub weight: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VectorPayloadRecord {
    pub user_id: String,
    pub kind: String,
    pub ref_id: String,
    pub text: String,
    pub tags: Vec<String>,
    pub source: String,
    pub content_hash: Option<String>,
    pub created_at: String,
    pub updated_at: Option<String>,
    #[serde(flatten)]
    pub metadata: HashMap<String, Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct QdrantPoint {
    pub id: Value,
    pub score: Option<f32>,
    pub payload: Option<Value>,
}

#[derive(Clone)]
pub struct QdrantClient {
    client: Client,
    config: AppConfig,
}

impl QdrantClient {
    pub fn new(client: Client, config: AppConfig) -> Self {
        Self { client, config }
    }

    pub async fn ensure_collection(&self, dimension: usize) -> AppResult<()> {
        let url = format!("{}/collections/{}", self.config.qdrant_url, self.config.qdrant_collection);
        let response = self.client.get(&url).send().await;
        let mut should_create = false;
        match response {
            Ok(response) if response.status().is_success() => {
                let json: Value = response
                    .json()
                    .await
                    .map_err(|error| AppError::internal(format!("Invalid Qdrant collection response: {error}")))?;
                let size = json
                    .get("result")
                    .and_then(|value| value.get("config"))
                    .and_then(|value| value.get("params"))
                    .and_then(|value| value.get("vectors"))
                    .and_then(|value| value.get(DENSE_VECTOR_NAME))
                    .and_then(|value| value.get("size"))
                    .and_then(Value::as_u64)
                    .unwrap_or_default() as usize;
                let has_sparse = json
                    .get("result")
                    .and_then(|value| value.get("config"))
                    .and_then(|value| value.get("params"))
                    .and_then(|value| value.get("sparse_vectors"))
                    .and_then(|value| value.get(SPARSE_VECTOR_NAME))
                    .is_some();
                if size != dimension || !has_sparse {
                    self.client
                        .delete(&url)
                        .send()
                        .await
                        .map_err(|error| AppError::internal(format!("Failed to delete Qdrant collection: {error}")))?;
                    should_create = true;
                }
            }
            _ => should_create = true,
        }

        if should_create {
            self.client
                .put(&url)
                .json(&json!({
                    "vectors": {
                        DENSE_VECTOR_NAME: { "size": dimension, "distance": "Cosine" }
                    },
                    "sparse_vectors": {
                        SPARSE_VECTOR_NAME: {}
                    }
                }))
                .send()
                .await
                .map_err(|error| AppError::internal(format!("Failed to create Qdrant collection: {error}")))?;
        }
        Ok(())
    }

    pub async fn upsert_points(&self, points: Vec<Value>) -> AppResult<()> {
        if points.is_empty() {
            return Ok(());
        }
        self.client
            .put(format!(
                "{}/collections/{}/points?wait=true",
                self.config.qdrant_url, self.config.qdrant_collection
            ))
            .json(&json!({ "points": points }))
            .send()
            .await
            .map_err(|error| AppError::internal(format!("Qdrant upsert failed: {error}")))?;
        Ok(())
    }

    pub async fn delete_points(&self, ids: Vec<Value>) -> AppResult<()> {
        if ids.is_empty() {
            return Ok(());
        }
        self.client
            .post(format!(
                "{}/collections/{}/points/delete?wait=true",
                self.config.qdrant_url, self.config.qdrant_collection
            ))
            .json(&json!({ "points": ids }))
            .send()
            .await
            .map_err(|error| AppError::internal(format!("Qdrant delete failed: {error}")))?;
        Ok(())
    }

    pub async fn query_dense_points(
        &self,
        vector: &[f32],
        user_id: &str,
        kinds: Option<&[String]>,
        limit: usize,
        score_threshold: Option<f32>,
    ) -> AppResult<Vec<QdrantPoint>> {
        self.query_points(json!(vector), DENSE_VECTOR_NAME, user_id, kinds, limit, score_threshold)
            .await
    }

    pub async fn search_hybrid_points(
        &self,
        dense_vector: &[f32],
        sparse_vector: &SparseVector,
        user_id: &str,
        kinds: Option<&[String]>,
        limit: usize,
        prefer_sparse: bool,
    ) -> AppResult<Vec<QdrantPoint>> {
        let candidate_limit = (limit * 3).clamp(12, 60);
        let dense = self
            .query_dense_points(dense_vector, user_id, kinds, candidate_limit, None)
            .await?;
        if sparse_vector.indices.is_empty() {
            return Ok(dense.into_iter().take(limit).collect());
        }
        let sparse = self
            .query_points(json!(sparse_vector), SPARSE_VECTOR_NAME, user_id, kinds, candidate_limit, None)
            .await?;
        Ok(fuse_reciprocal_rank(dense, sparse, prefer_sparse, limit))
    }

    pub async fn scroll_user_points(&self, user_id: &str) -> AppResult<Vec<QdrantPoint>> {
        let mut points = Vec::new();
        let mut offset = Value::Null;
        loop {
            let response = self
                .client
                .post(format!(
                    "{}/collections/{}/points/scroll",
                    self.config.qdrant_url, self.config.qdrant_collection
                ))
                .json(&json!({
                    "offset": offset,
                    "limit": 256,
                    "filter": create_filter(user_id, None),
                    "with_payload": true,
                    "with_vector": false
                }))
                .send()
                .await
                .map_err(|error| AppError::internal(format!("Qdrant scroll failed: {error}")))?;
            let json: Value = response
                .json()
                .await
                .map_err(|error| AppError::internal(format!("Invalid Qdrant scroll response: {error}")))?;
            let page: Vec<QdrantPoint> = serde_json::from_value(
                json.get("result")
                    .and_then(|value| value.get("points"))
                    .cloned()
                    .unwrap_or_else(|| json!([])),
            )
            .map_err(|error| AppError::internal(format!("Invalid Qdrant points: {error}")))?;
            let next_offset = json
                .get("result")
                .and_then(|value| value.get("next_page_offset"))
                .cloned()
                .unwrap_or(Value::Null);
            let done = next_offset.is_null() || page.is_empty();
            points.extend(page);
            if done {
                break;
            }
            offset = next_offset;
        }
        Ok(points)
    }

    async fn query_points(
        &self,
        query: Value,
        using: &str,
        user_id: &str,
        kinds: Option<&[String]>,
        limit: usize,
        score_threshold: Option<f32>,
    ) -> AppResult<Vec<QdrantPoint>> {
        let response = self
            .client
            .post(format!(
                "{}/collections/{}/points/query",
                self.config.qdrant_url, self.config.qdrant_collection
            ))
            .json(&json!({
                "query": query,
                "using": using,
                "limit": limit,
                "score_threshold": score_threshold,
                "filter": create_filter(user_id, kinds),
                "with_payload": true,
                "with_vector": false
            }))
            .send()
            .await
            .map_err(|error| AppError::internal(format!("Qdrant query failed: {error}")))?;
        let json: Value = response
            .json()
            .await
            .map_err(|error| AppError::internal(format!("Invalid Qdrant query response: {error}")))?;
        serde_json::from_value(
            json.get("result")
                .and_then(|value| value.get("points"))
                .cloned()
                .unwrap_or_else(|| json!([])),
        )
        .map_err(|error| AppError::internal(format!("Invalid Qdrant matches: {error}")))
    }
}

pub fn hash_point_id(user_id: &str, document_id: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(format!("{user_id}:{document_id}"));
    hex::encode(hasher.finalize())[..32].to_string()
}

pub fn hash_content(text: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(text);
    hex::encode(hasher.finalize())
}

pub fn build_sparse_vector(text: &str, boosts: &[SparseBoost]) -> SparseVector {
    let mut index_to_values: HashMap<u32, f32> = HashMap::new();
    for (token, weight) in tokenize_for_sparse_index(text) {
        let index = fnv1a32(&token) % SPARSE_HASH_SPACE;
        *index_to_values.entry(index).or_default() += weight;
    }
    for boost in boosts {
        let normalized = boost.token.trim().to_lowercase();
        if normalized.is_empty() || boost.weight <= 0.0 {
            continue;
        }
        let index = fnv1a32(&normalized) % SPARSE_HASH_SPACE;
        *index_to_values.entry(index).or_default() += boost.weight;
    }
    let mut ordered: Vec<(u32, f32)> = index_to_values.into_iter().collect();
    ordered.sort_by_key(|(index, _)| *index);
    SparseVector {
        indices: ordered.iter().map(|(index, _)| *index).collect(),
        values: ordered.iter().map(|(_, value)| (*value * 10_000.0).round() / 10_000.0).collect(),
    }
}

fn create_filter(user_id: &str, kinds: Option<&[String]>) -> Value {
    let mut must = vec![
        json!({ "key": "userId", "match": { "value": user_id } }),
        json!({ "key": "source", "match": { "value": RAG_VECTOR_SOURCE } }),
    ];
    if let Some(kinds) = kinds {
        if !kinds.is_empty() {
            must.push(json!({ "key": "kind", "match": { "any": kinds } }));
        }
    }
    json!({ "must": must })
}

fn tokenize_for_sparse_index(text: &str) -> Vec<(String, f32)> {
    let lowered = text.to_lowercase();
    let mut entries: HashMap<String, f32> = HashMap::new();
    let add_token = |entries: &mut HashMap<String, f32>, token: String, weight: f32| {
        if token.is_empty() {
            return;
        }
        *entries.entry(token).or_default() += weight;
    };

    let date_regex = regex::Regex::new(r"\b\d{4}-\d{2}(?:-\d{2})?\b|\b(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*\s+\d{4}\b").unwrap();
    for capture in date_regex.find_iter(&lowered) {
        add_token(&mut entries, capture.as_str().to_string(), 3.0);
    }

    let money_regex = regex::Regex::new(r"\$?\d+(?:\.\d+)?").unwrap();
    for capture in money_regex.find_iter(&lowered) {
        let raw = capture.as_str();
        let normalized = raw.trim_start_matches('$').to_string();
        let weight = if normalized.contains('.') || normalized.len() >= 4 || raw.contains('$') {
            2.5
        } else {
            1.2
        };
        add_token(&mut entries, normalized, weight);
        if raw.contains('$') {
            add_token(&mut entries, raw.to_string(), weight);
        }
    }

    let word_regex = regex::Regex::new(r"[a-z0-9][a-z0-9._:/-]{1,}").unwrap();
    for capture in word_regex.find_iter(&lowered) {
        let raw = capture.as_str();
        if raw.len() < 2 {
            continue;
        }
        let has_digit = raw.chars().any(|char| char.is_ascii_digit());
        add_token(&mut entries, raw.to_string(), if has_digit { 1.8 } else { 1.0 });
    }

    entries.into_iter().collect()
}

fn fnv1a32(input: &str) -> u32 {
    let mut hash: u32 = 0x811c9dc5;
    for byte in input.bytes() {
        hash ^= byte as u32;
        hash = hash.wrapping_mul(0x0100_0193);
    }
    hash
}

fn fuse_reciprocal_rank(
    dense: Vec<QdrantPoint>,
    sparse: Vec<QdrantPoint>,
    prefer_sparse: bool,
    limit: usize,
) -> Vec<QdrantPoint> {
    let dense_weight = if prefer_sparse { 0.4 } else { 0.6 };
    let sparse_weight = if prefer_sparse { 0.6 } else { 0.4 };
    let mut merged: HashMap<String, (QdrantPoint, f32)> = HashMap::new();
    for (index, point) in dense.into_iter().enumerate() {
        let key = point.id.to_string();
        let score = dense_weight / (RRF_K + index as f32 + 1.0);
        merged
            .entry(key)
            .and_modify(|(_, fused)| *fused += score)
            .or_insert((point, score));
    }
    for (index, point) in sparse.into_iter().enumerate() {
        let key = point.id.to_string();
        let score = sparse_weight / (RRF_K + index as f32 + 1.0);
        merged
            .entry(key)
            .and_modify(|(existing, fused)| {
                if existing.payload.is_none() {
                    existing.payload = point.payload.clone();
                }
                *fused += score;
            })
            .or_insert((point, score));
    }
    let mut values: Vec<(QdrantPoint, f32)> = merged.into_values().collect();
    values.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    values
        .into_iter()
        .take(limit)
        .map(|(mut point, score)| {
            point.score = Some(score);
            point
        })
        .collect()
}
